File: src/okbay/wiki.py
"""Markdown wiki pages with YAML-ish frontmatter and wikilinks."""
from __future__ import annotations
import re
from dataclasses import dataclass, field
from pathlib import Path

WIKILINK = re.compile(r"\[\[([^\]|#]+)(?:#[^\]|]+)?(?:\|[^\]]+)?\]\]")
FRONT = re.compile(r"\A---\n(.*?)\n---\n?", re.S)

def slugify(title: str) -> str:
    s = re.sub(r"[^a-z0-9]+", "-", title.strip().lower())
    return s.strip("-") or "page"

@dataclass
class Page:
    stem: str
    path: Path
    title: str
    kind: str = "note"
    body: str = ""
    front: dict = field(default_factory=dict)
    links: list[str] = field(default_factory=list)
    sources: list[str] = field(default_factory=list)
    def text(self) -> str:
        return f"{self.title}\n{self.body}"

def _parse_front(raw: str) -> dict:
    meta: dict = {}
    for line in raw.splitlines():
        if ":" not in line:
            continue
        k, v = line.split(":", 1)
        key = k.strip()
        val = v.strip().strip('"').strip("'")
        if val.startswith("[") and val.endswith("]"):
            inner = val[1:-1].strip()
            meta[key] = [p.strip().strip('"').strip("'") for p in inner.split(",") if p.strip()]
        else:
            meta[key] = val
    return meta

def parse_page(path: Path) -> Page | None:
    try:
        raw = path.read_text(encoding="utf-8")
    except OSError:
        return None
    stem = path.stem
    title = stem.replace("-", " ").title()
    kind = "note"
    body = raw
    front: dict = {}
    m = FRONT.match(raw)
    if m:
        front = _parse_front(m.group(1))
        body = raw[m.end():]
        title = str(front.get("title") or title)
        # CE pages often use type:; prefer kind, then type, for Atlas palette fidelity.
        kind = str(front.get("kind") or front.get("type") or kind)
        stem = str(front.get("stem") or stem)
    links = [g.strip() for g in WIKILINK.findall(body)]
    sources = front.get("sources") or front.get("extracted_from") or []
    if isinstance(sources, str):
        sources = [sources]
    return Page(stem=stem, path=path, title=title, kind=kind, body=body, front=front, links=links, sources=list(sources))

def list_pages(wiki_dir: Path | None = None) -> list[Page]:
    from . import paths
    wiki_dir = wiki_dir or paths.wiki()
    if not wiki_dir.exists():
        return []
    pages = []
    for p in sorted(wiki_dir.rglob("*.md")):
        page = parse_page(p)
        if page:
            pages.append(page)
    return pages

def write_page(wiki_dir: Path | None = None, stem: str = "", title: str = "", body: str = "", kind: str = "note", sources: list[str] | None = None, extra: dict | None = None, **kwargs) -> Page:
    from . import paths
    if wiki_dir is None:
        wiki_dir = paths.wiki()
    title = title or kwargs.get("title") or stem
    stem = stem or kwargs.get("stem") or title
    extra = extra or kwargs.get("extra")
    if extra and not sources:
        sources = extra.get("extracted_from") or extra.get("sources")
    wiki_dir.mkdir(parents=True, exist_ok=True)
    stem = slugify(stem or title)
    path = wiki_dir / f"{stem}.md"
    lines = ["---", f"stem: {stem}", f"title: {title}", f"kind: {kind}"]
    if sources:
        lines.append(f"extracted_from: [{', '.join(sources)}]")
    if extra:
        for k, v in extra.items():
            if isinstance(v, list):
                lines.append(f"{k}: [{', '.join(str(x) for x in v)}]")
            else:
                lines.append(f"{k}: {v}")
    lines.append("---")
    lines.append("")
    if not body.startswith("#"):
        lines.append(f"# {title}")
        lines.append("")
    lines.append(body.rstrip())
    lines.append("")
    path.write_text("\n".join(lines), encoding="utf-8")
    page = parse_page(path)
    if page is None:
        page = Page(stem=stem, path=path, title=title, kind=kind, body=body, front=extra or {}, links=[], sources=sources or [])
    return page

_STEM_INDEX: dict[str, Path] = {}
_STEM_INDEX_ROOT: Path | None = None
_STEM_INDEX_BUILT_AT: float = 0.0


def _wiki_mtime(directory: Path) -> float:
    try:
        return directory.stat().st_mtime
    except OSError:
        return 0.0


def _stem_index(directory: Path) -> dict[str, Path]:
    """Map file-stem → path. Built once per wiki root; avoids parsing 39k pages per modal open."""
    global _STEM_INDEX, _STEM_INDEX_ROOT, _STEM_INDEX_BUILT_AT
    mtime = _wiki_mtime(directory)
    if (
        _STEM_INDEX_ROOT == directory.resolve()
        and _STEM_INDEX
        and _STEM_INDEX_BUILT_AT >= mtime
    ):
        return _STEM_INDEX
    idx: dict[str, Path] = {}
    if directory.exists():
        for p in directory.rglob("*.md"):
            # Prefer first path for a stem; CE layout is wiki/<kind>/<stem>.md
            idx.setdefault(p.stem, p)
    _STEM_INDEX = idx
    try:
        _STEM_INDEX_ROOT = directory.resolve()
    except OSError:
        _STEM_INDEX_ROOT = directory
    _STEM_INDEX_BUILT_AT = mtime
    return idx



def get_page(stem: str, wiki_dir: Path | None = None) -> Page | None:
    from . import paths
    directory = wiki_dir or paths.wiki()
    if not stem:
        return None
    slug = slugify(stem)

    if not directory.exists():
        return None

    # 1) Flat wiki/<stem>.md
    direct = directory / f"{slug}.md"
    if direct.is_file():
        return parse_page(direct)

    # 2) Nested CE layout via stem index (one rglob to build, then O(1); no parse-all).
    idx = _stem_index(directory)
    hit = idx.get(slug) or idx.get(stem)
    if hit is not None and hit.is_file():
        return parse_page(hit)

    # 3) Last resort: single-name rglob (still no full parse).
    matches = list(directory.rglob(f"{slug}.md"))
    for p in matches:
        page = parse_page(p)
        if page is None:
            continue
        if page.stem == stem or page.stem == slug or page.title.lower() == stem.lower():
            return page
    if matches:
        return parse_page(matches[0])
    return None

Page.meta = property(lambda self: self.front)

def search_pages(query: str, limit: int = 20) -> list[dict]:
    from .store import search as fts_search
    from .graph import load, search_graph
    q = (query or "").strip()
    hits = fts_search(q, limit=limit) if q else fts_search("", limit=limit)
    if hits:
        return hits
    graph_hits = search_graph(load(), q)
    if graph_hits:
        return graph_hits[:limit]
    out = []
    needle = q.lower()
    for page in list_pages():
        blob = f"{page.stem} {page.title} {page.kind} {page.body}".lower()
        if needle in blob:
            out.append({"stem": page.stem, "title": page.title, "kind": page.kind, "path": str(page.path)})
        if len(out) >= limit:
            break
    return out

def neighbors(stem: str) -> dict:
    from .graph import load
    g = load()
    incoming, outgoing = [], []
    for e in g.get("edges") or []:
        if e.get("source") == stem:
            outgoing.append(e)
        if e.get("target") == stem:
            incoming.append(e)
    return {"stem": stem, "outgoing": outgoing, "incoming": incoming}


# ── Slice 2: lazy atlas page payload (do not embed bodies in /api/atlas/data) ─

_WIKILINK_MD = re.compile(r"\[\[([^\]|#]+)(?:#[^\]|]+)?(?:\|([^\]]+))?\]\]")
_MD_BOLD = re.compile(r"\*\*(.+?)\*\*")
_MD_ITALIC = re.compile(r"(?<!\*)\*(?!\*)(.+?)(?<!\*)\*(?!\*)")
_MD_CODE = re.compile(r"`([^`]+)`")
_MD_LINK = re.compile(r"\[([^\]]+)\]\(([^)]+)\)")


def _html_escape(s: str) -> str:
    return (
        str(s)
        .replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
    )


def markdown_to_html(md: str) -> str:
    """Minimal markdown→HTML for the slim atlas modal (no external deps)."""
    if not md:
        return ""

    def format_inline(raw: str) -> str:
        tokens: list[str] = []

        def protect(pattern: re.Pattern, repl_fn, s: str) -> str:
            out: list[str] = []
            last = 0
            for m in pattern.finditer(s):
                out.append(s[last : m.start()])
                ph = f"\x00{len(tokens)}\x00"
                tokens.append(repl_fn(m))
                out.append(ph)
                last = m.end()
            out.append(s[last:])
            return "".join(out)

        s = protect(
            _WIKILINK_MD,
            lambda m: (
                f'<a class="wikilink" data-page="{_html_escape(m.group(1).strip())}" '
                f'href="#page={_html_escape(m.group(1).strip())}">'
                f"{_html_escape((m.group(2) or m.group(1)).strip())}</a>"
            ),
            raw,
        )
        s = protect(
            _MD_LINK,
            lambda m: f'<a href="{_html_escape(m.group(2))}">{_html_escape(m.group(1))}</a>',
            s,
        )
        s = protect(_MD_CODE, lambda m: f"<code>{_html_escape(m.group(1))}</code>", s)
        s = _html_escape(s)
        for i, tok in enumerate(tokens):
            s = s.replace(f"\x00{i}\x00", tok)
        s = _MD_BOLD.sub(r"<strong>\1</strong>", s)
        s = _MD_ITALIC.sub(r"<em>\1</em>", s)
        return s

    lines = md.replace("\r\n", "\n").split("\n")
    out: list[str] = []
    in_ul = False
    in_code = False
    code_buf: list[str] = []

    def flush_ul() -> None:
        nonlocal in_ul
        if in_ul:
            out.append("</ul>")
            in_ul = False

    for line in lines:
        if line.strip().startswith("```"):
            if in_code:
                out.append("<pre><code>" + _html_escape("\n".join(code_buf)) + "</code></pre>")
                code_buf = []
                in_code = False
            else:
                flush_ul()
                in_code = True
            continue
        if in_code:
            code_buf.append(line)
            continue
        stripped = line.strip()
        if not stripped:
            flush_ul()
            continue
        if stripped.startswith("#"):
            flush_ul()
            level = len(stripped) - len(stripped.lstrip("#"))
            level = min(max(level, 1), 4)
            content = stripped[level:].strip()
            out.append(f"<h{level}>{format_inline(content)}</h{level}>")
            continue
        if stripped.startswith(("- ", "* ")):
            if not in_ul:
                out.append("<ul>")
                in_ul = True
            out.append(f"<li>{format_inline(stripped[2:].strip())}</li>")
            continue
        flush_ul()
        out.append(f"<p>{format_inline(stripped)}</p>")
    flush_ul()
    if in_code:
        out.append("<pre><code>" + _html_escape("\n".join(code_buf)) + "</code></pre>")
    return "\n".join(out)


def page_payload(stem: str, wiki_dir: Path | None = None) -> dict | None:
    """Lazy page doc for the atlas modal. Returns None if not found."""
    page = get_page(stem, wiki_dir=wiki_dir)
    if page is None:
        return None
    sources = list(page.sources or [])
    props: dict = {"sources": sources}
    for k, v in (page.front or {}).items():
        if k in ("stem", "title", "kind", "type", "sources", "extracted_from"):
            continue
        props[k] = v
    ntype = str(page.front.get("type") or page.kind or "note")
    return {
        "id": page.stem,
        "stem": page.stem,
        "title": page.title,
        "type": ntype,
        "kind": page.kind,
        "path": str(page.path),
        "properties": props,
        "sources": sources,
        "markdown": page.body,
        "body_html": markdown_to_html(page.body),
    }

File: crates/okbayd/src/wiki.rs
//! Markdown pages with YAML-ish frontmatter and [[wikilinks]].
use serde::Serialize;
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Clone, Debug, Serialize)]
pub struct Page {
    pub stem: String,
    pub path: PathBuf,
    pub title: String,
    pub kind: String,
    pub body: String,
    pub links: Vec<String>,
    pub sources: Vec<String>,
}

pub fn slugify(title: &str) -> String {
    let mut out = String::new();
    for ch in title.trim().to_lowercase().chars() {
        if ch.is_ascii_alphanumeric() { out.push(ch); }
        else if !out.ends_with('-') && !out.is_empty() { out.push('-'); }
    }
    let s = out.trim_matches('-').to_string();
    if s.is_empty() { "page".into() } else { s }
}

fn extract_wikilinks(body: &str) -> Vec<String> {
    let mut links = Vec::new();
    let bytes = body.as_bytes();
    let mut i = 0;
    while i + 1 < bytes.len() {
        if bytes[i] == b'[' && bytes[i + 1] == b'[' {
            if let Some(rel) = body[i + 2..].find("]]") {
                let inner = &body[i + 2..i + 2 + rel];
                let dest = inner.split('|').next().unwrap_or(inner);
                let dest = dest.split('#').next().unwrap_or(dest).trim();
                if !dest.is_empty() { links.push(dest.to_string()); }
                i += 4 + rel;
                continue;
            }
        }
        i += 1;
    }
    links
}

fn parse_front(raw: &str) -> std::collections::BTreeMap<String, String> {
    let mut meta = std::collections::BTreeMap::new();
    for line in raw.lines() {
        if let Some((k, v)) = line.split_once(':') {
            meta.insert(k.trim().to_string(), v.trim().trim_matches('"').trim_matches('\'').to_string());
        }
    }
    meta
}

fn parse_list(val: &str) -> Vec<String> {
    let t = val.trim();
    if t.starts_with('[') && t.ends_with(']') {
        t[1..t.len()-1].split(',').map(|p| p.trim().trim_matches('"').trim_matches('\'').to_string()).filter(|s| !s.is_empty()).collect()
    } else if t.is_empty() { vec![] } else { vec![t.to_string()] }
}

pub fn parse_page(path: &Path) -> Option<Page> {
    let raw = fs::read_to_string(path).ok()?;
    let stem_default = path.file_stem()?.to_string_lossy().to_string();
    let mut title = stem_default.replace('-', " ");
    let mut kind = "note".to_string();
    let mut stem = stem_default.clone();
    let mut body = raw.clone();
    let mut sources = vec![];
    if raw.starts_with("---\n") {
        if let Some(end) = raw[4..].find("\n---") {
            let front = parse_front(&raw[4..4 + end]);
            body = raw[4 + end + 4..].trim_start_matches('\n').to_string();
            if let Some(t) = front.get("title") { title = t.clone(); }
            // CE pages often use type:; prefer kind, then type, for Atlas palette fidelity.
            if let Some(k) = front.get("kind").or_else(|| front.get("type")) { kind = k.clone(); }
            if let Some(s) = front.get("stem") { stem = s.clone(); }
            if let Some(s) = front.get("sources").or_else(|| front.get("extracted_from")) { sources = parse_list(s); }
        }
    }
    Some(Page { stem, path: path.to_path_buf(), title, kind, body: body.clone(), links: extract_wikilinks(&body), sources })
}

pub fn list_pages(wiki_dir: &Path) -> Vec<Page> {
    let mut pages = vec![];
    let mut stack = vec![wiki_dir.to_path_buf()];
    while let Some(dir) = stack.pop() {
        let Ok(entries) = fs::read_dir(&dir) else { continue };
        for ent in entries.flatten() {
            let p = ent.path();
            if p.is_dir() { stack.push(p); }
            else if p.extension().and_then(|s| s.to_str()) == Some("md") {
                if let Some(page) = parse_page(&p) { pages.push(page); }
            }
        }
    }
    pages.sort_by(|a, b| a.stem.cmp(&b.stem));
    pages
}

pub fn get_page(wiki_dir: &Path, stem: &str) -> Option<Page> {
    list_pages(wiki_dir).into_iter().find(|p| p.stem == stem)
}

pub fn write_page(wiki_dir: &Path, stem: &str, title: &str, body: &str, kind: &str, sources: &[String]) -> Page {
    let _ = fs::create_dir_all(wiki_dir);
    let stem = slugify(if stem.is_empty() { title } else { stem });
    let path = wiki_dir.join(format!("{stem}.md"));
    let mut text = format!("---\nstem: {stem}\ntitle: {title}\nkind: {kind}\n");
    if !sources.is_empty() { text.push_str(&format!("extracted_from: [{}]\n", sources.join(", "))); }
    text.push_str("---\n\n");
    if !body.starts_with('#') { text.push_str(&format!("# {title}\n\n")); }
    text.push_str(body.trim_end());
    text.push('\n');
    let _ = fs::write(&path, text);
    parse_page(&path).unwrap_or(Page { stem, path, title: title.to_string(), kind: kind.to_string(), body: body.to_string(), links: vec![], sources: sources.to_vec() })
}


/// Slice 2: lazy atlas page payload (markdown + minimal HTML).
pub fn markdown_to_html(md: &str) -> String {
    if md.is_empty() { return String::new(); }
    fn esc(s: &str) -> String {
        s.replace('&', "&amp;").replace('<', "&lt;").replace('>', "&gt;").replace('"', "&quot;")
    }
    fn format_inline(raw: &str) -> String {
        // Very small: escape, then wikilinks, bold, italic, code.
        let mut out = String::new();
        let bytes = raw.as_bytes();
        let mut i = 0;
        while i < bytes.len() {
            if i + 1 < bytes.len() && bytes[i] == b'[' && bytes[i + 1] == b'[' {
                if let Some(rel) = raw[i + 2..].find("]]") {
                    let inner = &raw[i + 2..i + 2 + rel];
                    let mut parts = inner.splitn(2, '|');
                    let target = parts.next().unwrap_or("").split('#').next().unwrap_or("").trim();
                    let label = parts.next().unwrap_or(target).trim();
                    let tid = esc(target);
                    out.push_str(&format!(
                        "<a class=\"wikilink\" data-page=\"{tid}\" href=\"#page={tid}\">{}</a>",
                        esc(if label.is_empty() { target } else { label })
                    ));
                    i += 4 + rel;
                    continue;
                }
            }
            if bytes[i] == b'`' {
                if let Some(rel) = raw[i + 1..].find('`') {
                    out.push_str(&format!("<code>{}</code>", esc(&raw[i + 1..i + 1 + rel])));
                    i += 2 + rel;
                    continue;
                }
            }
            if i + 1 < bytes.len() && bytes[i] == b'*' && bytes[i + 1] == b'*' {
                if let Some(rel) = raw[i + 2..].find("**") {
                    out.push_str(&format!("<strong>{}</strong>", esc(&raw[i + 2..i + 2 + rel])));
                    i += 4 + rel;
                    continue;
                }
            }
            // single char escape
            match bytes[i] as char {
                '&' => out.push_str("&amp;"),
                '<' => out.push_str("&lt;"),
                '>' => out.push_str("&gt;"),
                '"' => out.push_str("&quot;"),
                c => out.push(c),
            }
            i += 1;
        }
        out
    }
    let mut out = String::new();
    let mut in_ul = false;
    for line in md.replace("\r\n", "\n").lines() {
        let stripped = line.trim();
        if stripped.is_empty() {
            if in_ul { out.push_str("</ul>"); in_ul = false; }
            continue;
        }
        if stripped.starts_with('#') {
            if in_ul { out.push_str("</ul>"); in_ul = false; }
            let level = stripped.chars().take_while(|c| *c == '#').count().clamp(1, 4);
            let content = stripped[level..].trim();
            out.push_str(&format!("<h{level}>{}</h{level}>", format_inline(content)));
            continue;
        }
        if stripped.starts_with("- ") || stripped.starts_with("* ") {
            if !in_ul { out.push_str("<ul>"); in_ul = true; }
            out.push_str(&format!("<li>{}</li>", format_inline(stripped[2..].trim())));
            continue;
        }
        if in_ul { out.push_str("</ul>"); in_ul = false; }
        out.push_str(&format!("<p>{}</p>", format_inline(stripped)));
    }
    if in_ul { out.push_str("</ul>"); }
    out
}

pub fn page_payload(wiki_dir: &Path, stem: &str) -> Option<serde_json::Value> {
    let page = get_page(wiki_dir, stem)?;
    let sources = page.sources.clone();
    Some(serde_json::json!({
        "id": page.stem,
        "stem": page.stem,
        "title": page.title,
        "type": page.kind,
        "kind": page.kind,
        "path": page.path.to_string_lossy(),
        "properties": { "sources": sources },
        "sources": page.sources,
        "markdown": page.body,
        "body_html": markdown_to_html(&page.body),
    }))
}
